### PR TITLE
Enable TruncationStrategy override for pipelines

### PR DESCRIPTION
**Pull request by Narsil**
# What does this PR do?

Right now truncation argument for tokenizers was not overridable, which leads to poor UX on some pipelines, most notably Summarization. Summaries trigger an error on text that end up with too many tokens for the underlying model.

Current strategy is just to enable the argument to be overrided as truncating by default is not necessarily good either.
More complex strategies are required to "solve" the problem (chunk original text into chunk of ~max_length, drop if some chunk is small enough <0.1 max_length?, then concatenate result summaries ?).

The current PR is a small step in that direction.
There should not be any backward incompatibilities with current changes.

@LysandreJik
@patrickvonplaten

## Before submitting
- [x] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/master/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [x] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/master/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/master/docs#writing-source-documentation).
- [x] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors which may be interested in your PR.

### CODE REVIEW

File: src/transformers/test.py [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+# Set graph optimization level
+sess_options.graph_optimization_level = rt.GraphOptimizationLevel.ORT_ENABLE_ALL
+
+sess = rt.InferenceSession("onnx/gpt2-optimized-quantized.onnx", sess_options)
+input_name = sess.get_inputs()[0].name
+X_test = np.zeros((1, 50))
+start = datetime.datetime.now()
+pred_onx = sess.run(None, {input_name: X_test.astype(np.long)})[0]
+print(datetime.datetime.now() - start)
+print(pred_onx)
```
Comment (LysandreJik) [inline src/transformers/test.py:19]: bogus file

File: tests/test_pipelines_common.py [hunk 2]
```diff
@@ -12,18 +12,86 @@
...
+from transformers.tokenization_utils_base import TruncationStrategy, to_py_obj
 
 
+if is_torch_available():
+    import torch
+
 VALID_INPUTS = ["A simple string", ["list of strings"]]
 
 
+class DummyTok:
```
Comment (LysandreJik) [inline tests/test_pipelines_common.py:31]: Why not use an existing tokenizer for this? Tokenizers are heavily tested, while this class looks does a lot (encoding, decoding, padding, truncation, converting to framework, creating attention masks, managing several types of inputs) and looks like it should be tested to ensure it behaves correctly.

**Review by LysandreJik — COMMENTED** [review]
Cool, this is a nice addition! 

Not too fond of the testing procedure though, I would rather we add tests based off of existing (tested) tokenizers, and test all the changed pipelines instead of just the summarization pipeline.

File: src/transformers/pipelines/text2text_generation.py [hunk 3]
```diff
@@ -96,7 +96,7 @@ def __call__(
             )
 
         with self.device_placement():
-            inputs = self._parse_and_tokenize(*documents, padding=padding)
+            inputs = self._parse_and_tokenize(*documents, padding=padding, **generate_kwargs)
```
Comment (patrickvonplaten) [inline src/transformers/pipelines/text2text_generation.py:99]: I think I'd prefer to either pop all tokenizer relevant tokens from `generate_kwargs` before this statement and only insert those or even directly add them as `keyword` arguments to the `__call__` method, like `padding_strategy=None` e.g.

File: src/transformers/pipelines/text2text_generation.py [hunk 4]
```diff
@@ -213,7 +216,7 @@ def __call__(
             )
 
         with self.device_placement():
-            inputs = self._parse_and_tokenize(*args, padding=padding)
+            inputs = self._parse_and_tokenize(*args, padding=padding, **generate_kwargs)
```
Comment (patrickvonplaten) [inline src/transformers/pipelines/text2text_generation.py:219]: same here

File: src/transformers/pipelines/text2text_generation.py [hunk 5]
```diff
@@ -320,11 +326,14 @@ def __call__(
...
```
Comment (patrickvonplaten) [inline src/transformers/pipelines/text2text_generation.py:329]: same here

File: src/transformers/pipelines/text_generation.py [hunk 6]
```diff
@@ -50,25 +50,15 @@ def __init__(self, *args, **kwargs):
...
-        inputs = self.tokenizer(
-            inputs,
-            add_special_tokens=add_special_tokens,
-            return_tensors=self.framework,
-            padding=padding,
-            **tokenizer_kwargs,
-        )
-
-        return inputs
+            kwargs.update({"add_space_before_punct_symbol": True})
```
Comment (patrickvonplaten) [inline src/transformers/pipelines/text_generation.py:59]: nice! I like this change

File: src/transformers/test.py [hunk 7]
```diff
@@ -0,0 +1,19 @@
+import datetime
```
Comment (patrickvonplaten) [inline src/transformers/test.py:1]: is this a left over file? Or you want to include this?

**Review by patrickvonplaten — APPROVED** [review]
Cool! Very welcoming change. My only nit is to not pass *generate_kwargs into a tokenizer as explained above. Great to have this tested now

File: tests/test_pipelines_common.py (hunk 2 above)
Comment (Narsil) [reply tests/test_pipelines_common.py:31]: The goal was to have an internet free tokenizer. (I was coding in the train so I really could not use internet there).

I could most definitely make a tokenizer from an existing class but that would require defining a vocabulary on the fly.

The biggest problem with a real tokenizer is the fact that the model is randomly initialized so guaranteeing the output is harder to do. Setting a seed might work, but I don't like to rely on that for tests as it tends to no work as good. Here I simply discard the actual output of the model and simply rely on the shapes.

I do agree that using a Dummy here might be overkill because we need to re implement a lot of functionality and we're not going to be sure that there's not a change between the two implementations.

Would:
- Using an in code vocabulary
- Setting a manual seed
be better in your opinion ?

File: src/transformers/pipelines/text2text_generation.py (hunk 3 above)
Comment (Narsil) [reply src/transformers/pipelines/text2text_generation.py:99]: I like that too, but that breaks backward compatibility right ? (If there was code written with an ignored kwarg, then we would start failing).

I could start something like: 

```python
__call__(self, tokenizer_arg1=XXX, tokenizer_arg2=XXXX, generate_kwarg1=XXX, ...., **kwargs):
  if kwargs:
      warnings.warn(UserWarning, "Unused keyword arguments {kwargs}" for pipeline {self}")

   self._parse_and_tokenizer(tokenizer_arg1=tokenizer_arg1, tokenizer_arg2=tokenizer_arg2)

   self.model.generate(generate_arg1, generate_kwarg2=generate_kwarg2, ...)
```
The only drawback of this, is we're calling `self.model.generate` over which I don't know if we control all kwargs. I mean are there model specific kwarg that we want to pass too (so we can't have an exhaustive list within `__call__` because we don't know ahead of time which arguments are available ?

File: src/transformers/test.py (hunk 7 above)
Comment (Narsil) [reply src/transformers/test.py:1]: It is bogus. I tried to untangle different changes from a branch of mine and this is what I get ;)

File: tests/test_pipelines_common.py (hunk 2 above)
Comment (LysandreJik) [reply tests/test_pipelines_common.py:31]: Hmm, we use internet for the tests so it wouldn't be an issue to have an internet-based tokenizer now that you're not in the train!

However, if you really want to go down the road of internet-free tokenizers, we do build some of them in the tokenization tests. Either by using the SentencePiece models in `tests/fixtures/` (see [here](https://github.com/huggingface/transformers/blob/master/tests/test_tokenization_albert.py#L41) an example for ALBERT's tokenizer), or by using an in-code vocabulary as you mention (see [here](https://github.com/huggingface/transformers/blob/master/tests/test_tokenization_bert.py#L44-L78) an example for BERT's tokenizer).

Regarding the randomly initialized model, you can also use a `tiny` variant of a model, which we create specifically for tests. There are already a bunch on the hub, and they sound less dangerous than setting a seed in a test that could potentially have side-effects on other tests down the road.

File: src/transformers/pipelines/text2text_generation.py (hunk 3 above)
Comment (patrickvonplaten) [reply src/transformers/pipelines/text2text_generation.py:99]: I don't understand why this would break backward compatibility. Previously no kwars were passed to `_parse_and_tokenize` => so the user had no control over the tokenizer at the `__call__` method of the pipeline. Now if we add kwargs, there should not be a problem in adding all possible tokenizer kwargs when correctly setting their defaults to the defaults in tokenizer no? But I see that there are a lot of tokenizer kwargs which don't all default to `None` :-/ Hmm, maybe the best solution is to just allow the `truncate` tokenizer argument for now and not the rest?

In general, as discussed with @LysandreJik and @mfuntowicz already multiple times, IMO the current pipelines design in Transformers is just not general enough to cleanly allow all use cases...

I think as it is now is quite dangerous if for some reason generate kwargs get the same name as tokenizer kwargs. 

It's definitely not an option to do `generate_kwags1` .... => there are way to many kwargs and maintainability becomes an issue?

File: src/transformers/pipelines/conversational.py [hunk 8]
```diff
@@ -273,12 +274,14 @@ def __call__(
             else:
                 return output
 
-    def _parse_and_tokenize(self, inputs, **kwargs):
+    def _parse_and_tokenize(
+        self, inputs, add_special_tokens=False, padding=False, truncation=TruncationStrategy.DO_NOT_TRUNCATE, **kwargs
```
Comment (patrickvonplaten) [inline src/transformers/pipelines/conversational.py:278]: Is kwargs just used to catch "unused" params?

File: src/transformers/pipelines/base.py [hunk 9]
```diff
@@ -577,7 +577,9 @@ def check_model_type(self, supported_models: Union[List[str], dict]):
                 f"The model '{self.model.__class__.__name__}' is not supported for {self.task}. Supported models are {supported_models}",
             )
 
-    def _parse_and_tokenize(self, inputs, padding=True, add_special_tokens=True, **kwargs):
+    def _parse_and_tokenize(
+        self, inputs, padding=True, add_special_tokens=True, truncation=TruncationStrategy.DO_NOT_TRUNCATE, **kwargs
```
Comment (patrickvonplaten) [inline src/transformers/pipelines/base.py:581]: Is "kwargs" just used to catch unused params here?

File: src/transformers/pipelines/text2text_generation.py (hunk 3 above)
Comment (patrickvonplaten) [reply src/transformers/pipelines/text2text_generation.py:99]: Actually, I just noticed that kwargs is not even really used in `_parse_and_tokenize` => then I think there is no problem in popping just `add_special_token` and `truncation` and not passing all the `generation_kwargs` -> why would this break anything?

Comment (Narsil) [reply src/transformers/pipelines/text2text_generation.py]: I changed that to follow your recommendation. Much better this way ! Thanks.

File: src/transformers/pipelines/base.py (hunk 9 above)
Comment (Narsil) [reply src/transformers/pipelines/base.py:581]: yes

File: src/transformers/pipelines/text2text_generation.py [hunk 10]
```diff
@@ -96,7 +107,7 @@ def __call__(
             )
 
         with self.device_placement():
-            inputs = self._parse_and_tokenize(*args, padding=padding, **generate_kwargs)
+            inputs = self._parse_and_tokenize(*args, padding=padding, truncation=truncation)
```
Comment (patrickvonplaten) [inline src/transformers/pipelines/text2text_generation.py:110]: Awesome

**Review by patrickvonplaten — APPROVED** [review]
LGTM!

**Review by LysandreJik — APPROVED** [review]
LGTM! Thanks for working on it @Narsil.
